### ISSUE TITLE
Enhancement: Ability to manage many-to-many associations with sorting support and etc.

### ISSUE THREAD

**Issue by romul** [smpallen99/ex_admin#96]
Hi,
I'd like to implement in ex_admin a support of complex cases with N:M associations and associations sorting support.
For example, you have categories, each category has many properties and each property could belongs to several categories. So, we have 3 tables: categories, properties, categories_properties (intersection table). 
Besides, we could set some extra attributes in categories_properties, like `position` (to sort properties inside a category), `required`, `displayable` and so on.

Admin interface could look like this:
![](https://api.monosnap.com/rpc/file/download?id=EWG5CL97FlrHRwdGYOlkDp3KTytI5i)

So, we have here:
1. a table with sortable rows (jquery-ui/sortable)
2. a field with autocomplete search of properties (select2)
3. some toggle-buttons to update boolean attributes without leaving the table
4. action-buttons like "delete" to remove association without leaving the table

I'm already working on this stuff, so some commits will be available soon. Are you interested in the described features?

**Comment by smpallen99** [comment]
Yes, I'm interested!! Couple notes:

1. I presume your already familiar with the sortable support as demonstrated in [Contact Demo](http://demo2.exadmin.info/admin/categories). 
2. Love the idea of autocomplete
3. I have some support for ajax support which i have not tested in a long time. I don't have any published examples or documentation on it yet. I have an old project that uses ajax form items to add additional fields based on a select box. It might give you some ideas for the ajax stuff...
4. Like an ajax delete button. Actually, we might want to consider ajax for all delete buttons at some time :)

I'm looking forward to this feature 👍

**Comment by smpallen99** [comment]
Here  is gist with some [dynamic form building code](https://gist.github.com/smpallen99/a65aebe5e80ded911c5b357a88e4d5b1) that used to work.

**Comment by romul** [comment]
@smpallen99 Thank you for the response.

Is there some documentation about JS & CSS organization inside ex_admin?
I guess, I didn't really get how active_admin_lib.js is being assembled and why static/themes is inside `priv` instead of `web` folder. So how could I add new js-file to active_admin_lib and where should be better to put some CSS rules?

**Comment by smpallen99** [comment]
@romul: To be honest its a real mess right now. I'm part way through a refactor that will clean much of it up. Probably a couple days away from finishing that work though. The following answers are based on the themes and master branches. Typically, I'm working in themes branch to develop new stuff. I then merge themes into master once i have pushed themes to githhub. Once I finish the static refactor, I'm going to create a new v0.7.4 branch and publish to hex.

With all that said... 

* After pulling the active_admin js from the active_admin project and finally getting it to build, I left the built stuff in priv. Unfortunately, I did not spend more time to create a repeatable build for it. This is why I have priv checked into git on the ex_admin project. 
* As part of the refactor to support themes, I just created a copy of all the active_admin static files and placed them under `priv/static/themes/active_admin`. I then built a similar structure under the new admin_lte2 theme.
* After my refactor, all the sources will be under `web/static/` somewhere. I was hoping to use `web/static/themes`, but I'm having trouble setting the babel ignore directive. So, I'm most likely going to put it all under `web/static/vendor/themes` unless there are common files between all the themes. 
* for the short term, you might want to create new files in `priv/static/themes` and include theme in the layout templates. When I get a little further on with the refactor, we can move them to a more appropriate place. 
* In the new admin_lte2 theme, I have a `ex_admin.css` file and an `ex_admin.js` file under `priv/static/themes/admin_lte2` that you can add too. 

Hope that helps, and sorry for the mess. I'm not very experienced with brunch and have been struggling with this part of the project. But It will get better...

**Comment by smpallen99** [comment]
BTW, my end goal is to have brunch build in the ex_admin project to build the installable js, css, and images under priv/static/themes, having 1 js and 1 css file under each of the themes. The admin.install will have an option for brunch. If no option, it will copy the files to projects `priv/static` directory. If brunch option, it will copy to projects `web/static/vendor` directory and provide instructions for editing `brunch-config.js`.

**Comment by romul** [comment]
@smpallen99 Glad to hear that improving of assets management is in progress :-)
I managed to make a first proof-of-concept implementation of managing associations position through ex_admin.
https://github.com/romul/synergy/blob/master/web/admin/category.ex#L18
https://github.com/romul/synergy/blob/master/web/controllers/admin/category_property_controller.ex
![](https://api.monosnap.com/rpc/file/download?id=tRTT8Iur5RslL6ZkItp2DXNaT8RohF)

Thinking about new macros like `sorted_table_for` and `sort_handle`
```
        sorted_table_for(Category.sorted_category_properties(category), "/admin/categories/#{category.id}/properties/update_positions") do
          sort_handle
          column :property
          column :required
          column :displayable
        end
```

**Comment by smpallen99** [comment]
Very cool!!

I wonder if we could handle the sorting by adding a sort option to the table_for macro. Also, would be nice if we could derive the ajax action so its implicit, and add option to override. I like the idea of using a collection_action and keeping it all in the resource file, instead of requiring a separate controller. 

I like the sort_handle helper.

Good work.

BTW, I just reviewed how I did the category sorting on demo2. Guess i missed the fact I had custom JS code that ended up in the package :( Something else to fix once you get this feature in...

**Comment by romul** [comment]
@smpallen99 
+1 for default implicit action, but I wonder why you are fond of putting all to one controller?
Honestly, it's already looks like a "god object". I guess, it would be much better to have several tiny controllers to improve readability of source code. We already have a `admin_routes` macro and separate namespace for admin controllers, so there is no problem to have many admin controllers.

**Comment by smpallen99** [comment]
@romul I get your point about "god object". So far, I've taken the approach that the resource file is where you go to customize the resource. Simply one file. I appreciate the push back on this concept. 

BTW, In the ajax example I posted yesterday, It had external controllers and views...

**Comment by romul** [comment]
Update: Now it works with the following syntax:
``` elixir
table_for(Category.sorted_category_properties(category), sortable: [resource: category, assoc_name: "category_properties"]) do
  sort_handle_column
  column :property
  column :required, toogle: true
  column :displayable, toogle: true
end
```

And looks like
![](https://api.monosnap.com/rpc/file/download?id=oeAcF9NVjYS7ql8NfHBjcNHKn0uLk1)

**Comment by romul** [comment]
@smpallen99 It's possible to change syntax from
`table_for(category.category_properties, sortable: [resource: category, assoc_name: "category_properties"]) do`
to
`sortable_table_for(category, :category_properties) do`

How do you feel, which option is better?

**Comment by smpallen99** [comment]
Yes, I'm good with `sortable_table_for`.

**Comment by smpallen99** [comment]
@romul Love the toggle! Will that be supported for any column, or just the table_for fields? I hope the former :)

What do you think of adding an option for the text on the toggle buttons? Something like
```
column :notification, toggle: ~w(On Off)
```

**Comment by romul** [comment]
> column :notification, toggle: ~w(On Off)

Excellent idea! Will do.

> Will that be supported for any column, or just the table_for fields?

I have also tested with `index`.. and it works as well:

``` elixir
    index do
      selectable_column
      column :category
      column :property
      column :required, toogle: true
      column :displayable, toogle: true
    end
```

**Comment by romul** [comment]
@smpallen99 Done:
```elixir
sortable_table_for(category, :category_properties) do
  sort_handle_column
  column :property
  column :required, toogle: ~w(yup nop)
  column :displayable, toogle: true
end
```
I like this syntax very much.

So b/c this task is fairly large, I propose the following next steps:

1. I'll polish these 2 features and add support of them to active_admin theme as well.
2. You will review and merge it to master
3. Some fixes based on code review if required
4. I'll continue work on AJAX adding associations through select2 and AJAX deletion in separate issues/tasks.

Deal?

**Comment by smpallen99** [comment]
@romul Would really like it if you can submit pull request on my ecto-1.1 branch. I've been doing all my dev on that branch and then merging to master. Does that work for you?

**Comment by romul** [comment]
@smpallen99 Ok

**Comment by smpallen99** [comment]
Thanks @romul for a great enhancement!! I've merged it into ecto-1.1 and master.

**Comment by romul** [comment]
👍
